# COVID-19 - Clinical Data to assess diagnosis

# Lesson 01 - Health and Machine Learning

Desafio 01: Se você não conhece o Kaggle, explore um pouco os datasets, competições e discussões para verificar o quão rico é esta plataforma.

Desafio 02: Leia a descrição do problema proposta pelo Sírio Libanês, tentando entender qual é o real problema. Antes de continuar para as próximas aulas, debata no Discord suas conclusões sobre o projeto.

# Lesson 02 - Kaggle


Firsts impressions from the data

In [1]:
from zipfile import ZipFile
from io import BytesIO
import requests
import pandas as pd
# Open data from github zip file
# get the file by requesting the url

In [3]:
data_url_zip = requests.get('https://github.com/gabrielatrindade/alura-bootcamp/blob/master/module5/data/patient.zip?raw=true')

zipfile = ZipFile(BytesIO(data_url_zip.content)) # convert data from get into binary data
file_name= zipfile.namelist()[0] # return file name from zifile namlist content

patients = pd.read_excel(zipfile.open(file_name).read())

patients.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                         0            1          60th       0   
1                         0            1          60th       0   
2                         0            1          60th       0   
3                         0            1          60th       0   
4                         0            1          60th       0   

   DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                 0.0                 0.0                 0.0   
1                 0.0                 0.0                 0.0   
2                 0.0                 0.0                 0.0   
3                 0.0                 0.0                 0.0   
4                 0.0                 0.0                 0.0   

   DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                 0.0                 1.0                 1.0  ...   
1                 0.0                 1.0                 1.0  ...   
2                 0.0                 1.0                 1.0  ...   
3                 0.0                 1.0                 1.0  ...   
4                 0.0                 1.0                 1.0  ...   

   TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  BLOODPRESSURE_DIASTOLIC_DIFF_REL  \
0         -1.000000               -1.000000                         -1.000000   
1         -1.000000               -1.000000                         -1.000000   
2               NaN                     NaN                               NaN   
3         -1.000000               -1.000000                               NaN   
4         -0.238095               -0.818182                         -0.389967   

   BLOODPRESSURE_SISTOLIC_DIFF_REL  HEART_RATE_DIFF_REL  \
0                        -1.000000            -1.000000   
1                        -1.000000            -1.000000   
2                              NaN                  NaN   
3                              NaN                  NaN   
4                         0.407558            -0.230462   

   RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
0                  -1.000000             -1.000000   
1                  -1.000000             -1.000000   
2                        NaN                   NaN   
3                        NaN             -1.000000   
4                   0.096774             -0.242282   

   OXYGEN_SATURATION_DIFF_REL    WINDOW  ICU  
0                   -1.000000       0-2    0  
1                   -1.000000       2-4    0  
2                         NaN       4-6    0  
3                   -1.000000      6-12    0  
4                   -0.814433  ABOVE_12    1  

[5 rows x 231 columns]

In [4]:
patients.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65       GENDER  DISEASE GROUPING 1  \
count               1925.000000  1925.000000  1925.000000         1920.000000   
mean                 192.000000     0.467532     0.368831            0.108333   
std                  111.168431     0.499074     0.482613            0.310882   
min                    0.000000     0.000000     0.000000            0.000000   
25%                   96.000000     0.000000     0.000000            0.000000   
50%                  192.000000     0.000000     0.000000            0.000000   
75%                  288.000000     1.000000     1.000000            0.000000   
max                  384.000000     1.000000     1.000000            1.000000   

       DISEASE GROUPING 2  DISEASE GROUPING 3  DISEASE GROUPING 4  \
count         1920.000000         1920.000000         1920.000000   
mean             0.028125            0.097917            0.019792   
std              0.165373            0.297279            0.139320   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE GROUPING 5  DISEASE GROUPING 6          HTN  ...  \
count         1920.000000         1920.000000  1920.000000  ...   
mean             0.128125            0.046875     0.213021  ...   
std              0.334316            0.211426     0.409549  ...   
min              0.000000            0.000000     0.000000  ...   
25%              0.000000            0.000000     0.000000  ...   
50%              0.000000            0.000000     0.000000  ...   
75%              0.000000            0.000000     0.000000  ...   
max              1.000000            1.000000     1.000000  ...   

       RESPIRATORY_RATE_DIFF  TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
count            1177.000000       1231.000000             1239.000000   
mean               -0.703683         -0.770338               -0.887196   
std                 0.482097          0.319001                0.296147   
min                -1.000000         -1.000000               -1.000000   
25%                -1.000000         -1.000000               -1.000000   
50%                -1.000000         -0.976190               -0.979798   
75%                -0.647059         -0.595238               -0.878788   
max                 1.000000          1.000000                1.000000   

       BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
count                       1240.000000                      1240.000000   
mean                          -0.786997                        -0.715950   
std                            0.324754                         0.419103   
min                           -1.000000                        -1.000000   
25%                           -1.000000                        -1.000000   
50%                           -1.000000                        -0.984944   
75%                           -0.645482                        -0.522176   
max                            1.000000                         1.000000   

       HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
count          1240.000000                1177.000000           1231.000000   
mean             -0.817800                  -0.719147             -0.771327   
std               0.270217                   0.446600              0.317694   
min              -1.000000                  -1.000000             -1.000000   
25%              -1.000000                  -1.000000             -1.000000   
50%              -0.989822                  -1.000000             -0.975924   
75%              -0.662529                  -0.634409             -0.594677   
max               1.000000                   1.000000              

In [5]:
patients['ICU']

0       0
1       0
2       0
3       0
4       1
       ..
1920    0
1921    0
1922    0
1923    0
1924    0
Name: ICU, Length: 1925, dtype: int64

In [6]:
patients['ICU'].value_counts()

0    1410
1     515
Name: ICU, dtype: int64

### Challenge 01: Study and Learn about supervised and unsupervised learnings, reinforcement and etc.
From wwcode slides:

- Supervised Learning: The algorithm learns from example data and associated target responses, in order to later predict the correct response when posed with new examples comes. <br>
    This approach is similar to a human learning under the supervision of a teacher.
    
    
- Unsupervised Learning: The algorithm learns from example data without any associated response, leaving to the algorithm to determine the data patterns on its own. <br>
    This is similar to methods humans use to figure out that certain objects or events are from the same class by observing the degree of similarity between objects.
    
    
- Semi-supervised Learning: Class of algorithms that are able to learn from partially labeled data sets. <br>
    Supervised Learning requires the dataset to be labeled either - this is a very costly process, especially when dealing with large volumes of data. <br>
    Unsupervised Learning has the disadvantage of limited application spectrum.
    
    
- Reinforcement Learning: The algorithm learns from example data along with positive/negative feedback to each prediction the algorithm proposes. <br>
    The algorithm must make decisions and the decisions bear “consequences”. <br>
    In the human world, it is just like learning by trial and error.

### Challenge 02: Study and Learn about regression and classification, etc.

▸ Classification - for categorical data
<br>
▸ Regression - for continuous data


# Lesson 03 - Machine Learning model

Creating the first Machine Learning model

In the wrong way! :p

In [7]:
from sklearn.linear_model import LogisticRegression

In [18]:
patients_useful = patients.dropna()

In [19]:
target_patients = patients_useful['ICU'].copy()
features_patients = patients_useful.drop('ICU', axis=1).copy()

features_patients.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
4                          0            1          60th       0   
6                          1            1          90th       1   
8                          1            1          90th       1   
9                          1            1          90th       1   
14                         2            0          10th       0   

    DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
4                  0.0                 0.0                 0.0   
6                  0.0                 0.0                 0.0   
8                  0.0                 0.0                 0.0   
9                  0.0                 0.0                 0.0   
14                 0.0                 0.0                 0.0   

    DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
4                  0.0                 1.0                 1.0  ...   
6                  0.0                 0.0                 0.0  ...   
8                  0.0                 0.0                 0.0  ...   
9                  0.0                 1.0                 0.0  ...   
14                 0.0                 0.0                 0.0  ...   

    RESPIRATORY_RATE_DIFF  TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
4                0.176471         -0.238095               -0.818182   
6               -1.000000         -1.000000               -1.000000   
8               -0.764706         -0.880952               -1.000000   
9                0.764706          0.142857               -0.797980   
14              -0.705882         -0.500000               -0.898990   

    BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
4                          -0.389967                         0.407558   
6                          -1.000000                        -1.000000   
8                          -0.906832                        -0.831132   
9                           0.315690                         0.200359   
14                         -0.612422                        -0.343258   

    HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
4             -0.230462                   0.096774             -0.242282   
6             -1.000000                  -1.000000             -1.000000   
8             -0.940967                  -0.817204             -0.882574   
9             -0.239515                   0.645161              0.139709   
14            -0.576744                  -0.695341             -0.505464   

    OXYGEN_SATURATION_DIFF_REL    WINDOW  
4                    -0.814433  ABOVE_12  
6                    -1.000000       2-4  
8                    -1.000000      6-12  
9                    -0.802317  ABOVE_12  
14                   -0.900129  ABOVE_12  

[5 rows x 230 columns]

The model works only with int or float columns.

In [20]:
patients_useful.describe().columns

Index(['PATIENT_VISIT_IDENTIFIER', 'AGE_ABOVE65', 'GENDER',
       'DISEASE GROUPING 1', 'DISEASE GROUPING 2', 'DISEASE GROUPING 3',
       'DISEASE GROUPING 4', 'DISEASE GROUPING 5', 'DISEASE GROUPING 6', 'HTN',
       ...
       'RESPIRATORY_RATE_DIFF', 'TEMPERATURE_DIFF', 'OXYGEN_SATURATION_DIFF',
       'BLOODPRESSURE_DIASTOLIC_DIFF_REL', 'BLOODPRESSURE_SISTOLIC_DIFF_REL',
       'HEART_RATE_DIFF_REL', 'RESPIRATORY_RATE_DIFF_REL',
       'TEMPERATURE_DIFF_REL', 'OXYGEN_SATURATION_DIFF_REL', 'ICU'],
      dtype='object', length=229)

In [21]:
x_columns = patients_useful.describe().columns
features_patients = patients_useful[x_columns].drop('ICU', axis=1)

In [22]:
LR_model = LogisticRegression()

LR_model.fit(features_patients, target_patients)

/home/gabriela/.local/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression()

**Challenge 01: Read about ethics in the data area, IA.**

In [24]:
features_patients.iloc[4]

PATIENT_VISIT_IDENTIFIER           2.000000
AGE_ABOVE65                        0.000000
GENDER                             0.000000
DISEASE GROUPING 1                 0.000000
DISEASE GROUPING 2                 0.000000
                                     ...   
BLOODPRESSURE_SISTOLIC_DIFF_REL   -0.343258
HEART_RATE_DIFF_REL               -0.576744
RESPIRATORY_RATE_DIFF_REL         -0.695341
TEMPERATURE_DIFF_REL              -0.505464
OXYGEN_SATURATION_DIFF_REL        -0.900129
Name: 14, Length: 228, dtype: float64

In [26]:
LR_model.predict([features_patients.iloc[4]])

array([1])

In [27]:
target_patients.iloc[4]

1

In [32]:
(LR_model.predict(features_patients) == target_patients).value_counts()

True     497
False     48
Name: ICU, dtype: int64

In [38]:
accuracy = sum(LR_model.predict(features_patients) == target_patients) / len(features_patients) * 100
accuracy

91.19266055045871

In [37]:
sum(LR_model.predict(features_patients) != target_patients) / len(features_patients) * 100

8.807339449541285

### But, of course, it doenst make sense train and predict/test with the same sample.

**Challenge 02: Why did we drop the target from features to fit the model? What will happen if we fit the model using a target inside the features (X)?**

Note: Logistic Regression 'transform a regression into classification'. For example, when I think about a someone who bet the lottery. I can say that the probability of him winning is very small.. or that it is zero (converting regression problem into classification one).

### Challenge 01: Read about ethics in the data area, IA.

Read https://towardsdatascience.com/the-ethics-of-data-science-e3b1828affa2

https://www.amazon.com/Ethical-Algorithm-Science-Socially-Design/dp/0190948205



### Challenge 02: Why did we drop the target from features to fit the model? What will happen if we fit the model using a target inside the features (X)?

We need to show what our target is and what is the features that we will use to train the model. The match of the target and features values will give to the model the ability to learn from the data.

# Lesson 04 - Metrics and Evaluation

Measuring and evaluating the model

In [40]:
target_patients.value_counts()

0    281
1    264
Name: ICU, dtype: int64

In [104]:
from sklearn.dummy import DummyClassifier

from sklearn.metrics import accuracy_score

from sklearn.model_selection import train_test_split

import numpy as np
np.random.seed(73246)

https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyClassifier.html

strategy
“prior”: always predicts the class that maximizes the class prior (like “most_frequent”) and predict_proba returns the class prior.

In [105]:
dummy_model = DummyClassifier()

dummy_model.fit(features_patients, target_patients)

sum(dummy_model.predict(features_patients) == target_patients) / len(features_patients) * 100

/home/gabriela/.local/lib/python3.6/site-packages/sklearn/dummy.py:132: FutureWarning: The default value of strategy will change from stratified to prior in 0.24.
  "stratified to prior in 0.24.", FutureWarning)


46.788990825688074

In [106]:
accuracy_score(target_patients, dummy_model.predict(features_patients)) 

0.5192660550458715

In [107]:
x_train, x_test, y_train, y_test = train_test_split(features_patients, 
                                                    target_patients, 
                                                    stratify=target_patients)

# data is split in a stratified fashion, using this as the class labels.
# that means, the data for train and test will be balanced in the target.
# if the x is split by 90% of the target is 0 and 10 is 1, 
# so the target should have the same proportion

In [108]:
print(len(x_train), len(y_train))
print(len(x_test), len(y_test))

408 408
137 137


In [109]:
model_LR = LogisticRegression()

model_LR.fit(x_train, y_train)

y_pred = model_LR.predict(x_test)
accuracy_score(y_test, y_pred)

/home/gabriela/.local/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.8978102189781022

let's see how it was split with `stratify` defined.

In [110]:
y_train.value_counts(normalize=True)

0    0.514706
1    0.485294
Name: ICU, dtype: float64

In [111]:
y_test.value_counts(normalize=True)

0    0.518248
1    0.481752
Name: ICU, dtype: float64

### Challenge 01: See a parameter that defines how much you want the test sample size and train sample size. What is the rule for it? How choose these number?

`test_size` or `train_size` parameters represent the proportion of the dataset to include in the test or train size, respectively.

https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

train test split sizes:

https://data36.com/predictive-analytics-101-part-1/

https://stackoverflow.com/questions/13610074/is-there-a-rule-of-thumb-for-how-to-divide-a-dataset-into-training-and-validatio/13623707#13623707

https://www.researchgate.net/post/Is-there-an-ideal-ratio-between-a-training-set-and-validation-set-Which-trade-off-would-you-suggest

### Challenge 02: See Logistic Regression documentation... What are the common parameters between LR and train test split? What does it means?

`random_state` is a common parameter between them.

Logistic Regression: https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html



# Lesson 05 - Understanding the real problem

Manipulating data for the right model.

In [115]:
patients.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                         0            1          60th       0   
1                         0            1          60th       0   
2                         0            1          60th       0   
3                         0            1          60th       0   
4                         0            1          60th       0   

   DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                 0.0                 0.0                 0.0   
1                 0.0                 0.0                 0.0   
2                 0.0                 0.0                 0.0   
3                 0.0                 0.0                 0.0   
4                 0.0                 0.0                 0.0   

   DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                 0.0                 1.0                 1.0  ...   
1                 0.0                 1.0                 1.0  ...   
2                 0.0                 1.0                 1.0  ...   
3                 0.0                 1.0                 1.0  ...   
4                 0.0                 1.0                 1.0  ...   

   TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  BLOODPRESSURE_DIASTOLIC_DIFF_REL  \
0         -1.000000               -1.000000                         -1.000000   
1         -1.000000               -1.000000                         -1.000000   
2               NaN                     NaN                               NaN   
3         -1.000000               -1.000000                               NaN   
4         -0.238095               -0.818182                         -0.389967   

   BLOODPRESSURE_SISTOLIC_DIFF_REL  HEART_RATE_DIFF_REL  \
0                        -1.000000            -1.000000   
1                        -1.000000            -1.000000   
2                              NaN                  NaN   
3                              NaN                  NaN   
4                         0.407558            -0.230462   

   RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
0                  -1.000000             -1.000000   
1                  -1.000000             -1.000000   
2                        NaN                   NaN   
3                        NaN             -1.000000   
4                   0.096774             -0.242282   

   OXYGEN_SATURATION_DIFF_REL    WINDOW  ICU  
0                   -1.000000       0-2    0  
1                   -1.000000       2-4    0  
2                         NaN       4-6    0  
3                   -1.000000      6-12    0  
4                   -0.814433  ABOVE_12    1  

[5 rows x 231 columns]

PATIENT_VISIT_IDENTIFIER is duplicated, which means that each row is not a unique patient, it's about the patient status.

In [116]:
patients.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65       GENDER  DISEASE GROUPING 1  \
count               1925.000000  1925.000000  1925.000000         1920.000000   
mean                 192.000000     0.467532     0.368831            0.108333   
std                  111.168431     0.499074     0.482613            0.310882   
min                    0.000000     0.000000     0.000000            0.000000   
25%                   96.000000     0.000000     0.000000            0.000000   
50%                  192.000000     0.000000     0.000000            0.000000   
75%                  288.000000     1.000000     1.000000            0.000000   
max                  384.000000     1.000000     1.000000            1.000000   

       DISEASE GROUPING 2  DISEASE GROUPING 3  DISEASE GROUPING 4  \
count         1920.000000         1920.000000         1920.000000   
mean             0.028125            0.097917            0.019792   
std              0.165373            0.297279            0.139320   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE GROUPING 5  DISEASE GROUPING 6          HTN  ...  \
count         1920.000000         1920.000000  1920.000000  ...   
mean             0.128125            0.046875     0.213021  ...   
std              0.334316            0.211426     0.409549  ...   
min              0.000000            0.000000     0.000000  ...   
25%              0.000000            0.000000     0.000000  ...   
50%              0.000000            0.000000     0.000000  ...   
75%              0.000000            0.000000     0.000000  ...   
max              1.000000            1.000000     1.000000  ...   

       RESPIRATORY_RATE_DIFF  TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
count            1177.000000       1231.000000             1239.000000   
mean               -0.703683         -0.770338               -0.887196   
std                 0.482097          0.319001                0.296147   
min                -1.000000         -1.000000               -1.000000   
25%                -1.000000         -1.000000               -1.000000   
50%                -1.000000         -0.976190               -0.979798   
75%                -0.647059         -0.595238               -0.878788   
max                 1.000000          1.000000                1.000000   

       BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
count                       1240.000000                      1240.000000   
mean                          -0.786997                        -0.715950   
std                            0.324754                         0.419103   
min                           -1.000000                        -1.000000   
25%                           -1.000000                        -1.000000   
50%                           -1.000000                        -0.984944   
75%                           -0.645482                        -0.522176   
max                            1.000000                         1.000000   

       HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
count          1240.000000                1177.000000           1231.000000   
mean             -0.817800                  -0.719147             -0.771327   
std               0.270217                   0.446600              0.317694   
min              -1.000000                  -1.000000             -1.000000   
25%              -1.000000                  -1.000000             -1.000000   
50%              -0.989822                  -1.000000             -0.975924   
75%              -0.662529                  -0.634409             -0.594677   
max               1.000000                   1.000000              

we have patients from 0 to 384.


Note: each row has a `window`, that means, the measurements - blood pressure, temperature - were taken in that time window. When the `ICU` is 1, we cannot know whether these measurements were taken in the moment, before or after the patient went to the ICU. So, we cannot considerate the rows with `ICU` equals 1.

Try to solve it: We can considerate the line with the `ICU` equals 1. We try predict if the next window the patient will go to the ICU or not from the previous window information, which the `ICU` is 0.

There is another way to solve that.


thinking about fill nan values... as the measurements are taking in short period of time, we can considerate to fill NaN values with the value before or after them.

In [127]:
fake_data = pd.DataFrame([[1,1, 1], [1, None, 2], 
                          [2, None, 1], [2, None, 2]],
                         columns=['user','measurement', 'window'])
fake_data

user  measurement  window
0     1          1.0       1
1     1          NaN       2
2     2          NaN       1
3     2          NaN       2

In [128]:
fake_data.fillna(method='ffill')

user  measurement  window
0     1          1.0       1
1     1          1.0       2
2     2          1.0       1
3     2          1.0       2

the patient 2 None information was filled by patient 1 information.

In [135]:
fake_data.groupby(['user'], as_index=False)['measurement'].fillna(method='ffill')

measurement
0          1.0
1          1.0
2          NaN
3          NaN

In [143]:
fake_data = pd.DataFrame([[1,None, 1], [1, 2.5, 2], 
                          [2, 3.0, 1], [2, None, 2],
                          [3, None, 1], [3, 1, 2], [3, None, 3],
                          [4, None, 1], [4, 6, 2], [4, None, 3], [4, 5, 4], [4, None, 5]],
                         columns=['user','measurement', 'window'])

fake_data

user  measurement  window
0      1          NaN       1
1      1          2.5       2
2      2          3.0       1
3      2          NaN       2
4      3          NaN       1
5      3          1.0       2
6      3          NaN       3
7      4          NaN       1
8      4          6.0       2
9      4          NaN       3
10     4          5.0       4
11     4          NaN       5

In [144]:
fake_data.groupby(['user'], as_index=False)['measurement'].fillna(method='ffill').fillna(method='bfill')

measurement
0           2.5
1           2.5
2           3.0
3           3.0
4           1.0
5           1.0
6           1.0
7           6.0
8           6.0
9           6.0
10          5.0
11          5.0

In [399]:
continuous_features_cols = patients.iloc[:, 13:-2].columns

continuous_features_cols

Index(['ALBUMIN_MEDIAN', 'ALBUMIN_MEAN', 'ALBUMIN_MIN', 'ALBUMIN_MAX',
       'ALBUMIN_DIFF', 'BE_ARTERIAL_MEDIAN', 'BE_ARTERIAL_MEAN',
       'BE_ARTERIAL_MIN', 'BE_ARTERIAL_MAX', 'BE_ARTERIAL_DIFF',
       ...
       'HEART_RATE_DIFF', 'RESPIRATORY_RATE_DIFF', 'TEMPERATURE_DIFF',
       'OXYGEN_SATURATION_DIFF', 'BLOODPRESSURE_DIASTOLIC_DIFF_REL',
       'BLOODPRESSURE_SISTOLIC_DIFF_REL', 'HEART_RATE_DIFF_REL',
       'RESPIRATORY_RATE_DIFF_REL', 'TEMPERATURE_DIFF_REL',
       'OXYGEN_SATURATION_DIFF_REL'],
      dtype='object', length=216)

In [400]:
def fill_df_continuous_vars(df: pd.DataFrame):
    
    continuous_features_cols = df.iloc[:, 13:-2].columns

    filled_continuous_features = (df.groupby('PATIENT_VISIT_IDENTIFIER')
                                  [continuous_features_cols]
                                  .fillna(method='bfill')
                                  .fillna(method='ffill')
                                  .reset_index())
    
    categorical_features = df.iloc[:, :13]
    
    output = df.iloc[:, -2:]
    
    return pd.concat([categorical_features, filled_continuous_features, output], axis=1)

In [401]:
patients_filled_continuous_vars = fill_df_continuous_vars(patients)

patients_filled_continuous_vars

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                            0            1          60th       0   
1                            0            1          60th       0   
2                            0            1          60th       0   
3                            0            1          60th       0   
4                            0            1          60th       0   
...                        ...          ...           ...     ...   
1920                       384            0          50th       1   
1921                       384            0          50th       1   
1922                       384            0          50th       1   
1923                       384            0          50th       1   
1924                       384            0          50th       1   

      DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                    0.0                 0.0                 0.0   
1                    0.0                 0.0                 0.0   
2                    0.0                 0.0                 0.0   
3                    0.0                 0.0                 0.0   
4                    0.0                 0.0                 0.0   
...                  ...                 ...                 ...   
1920                 0.0                 0.0                 0.0   
1921                 0.0                 0.0                 0.0   
1922                 0.0                 0.0                 0.0   
1923                 0.0                 0.0                 0.0   
1924                 0.0                 0.0                 1.0   

      DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                    0.0                 1.0                 1.0  ...   
1                    0.0                 1.0                 1.0  ...   
2                    0.0                 1.0                 1.0  ...   
3                    0.0                 1.0                 1.0  ...   
4                    0.0                 1.0                 1.0  ...   
...                  ...                 ...                 ...  ...   
1920                 0.0                 0.0                 0.0  ...   
1921                 0.0                 0.0                 0.0  ...   
1922                 0.0                 0.0                 0.0  ...   
1923                 0.0                 0.0                 0.0  ...   
1924                 0.0                 0.0                 0.0  ...   

      TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
0            -1.000000               -1.000000   
1            -1.000000               -1.000000   
2            -1.000000               -1.000000   
3            -1.000000               -1.000000   
4            -0.238095               -0.818182   
...                ...                     ...   
1920         -1.000000               -1.000000   
1921         -1.000000               -1.000000   
1922         -1.000000               -1.000000   
1923         -1.000000               -1.000000   
1924         -0.547619               -0.838384   

      BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
0                            -1.000000                        -1.000000   
1                            -1.000000                        -1.000000   
2                            -0.389967                         0.407558   
3                            -0.389967                         0.407558   
4                            -0.389967                         0.407558   
...                                ...                              ...   
1920                         -1.000000                        -1.000000   
1921                         -1.000000                        -1.000000   
1922                         -1.000000                        -1.000000   
1923                         -1.000000                        -1.000000   
1924                         -0.701863                   

In [402]:
patients_filled_continuous_var.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65       GENDER  DISEASE GROUPING 1  \
count               1925.000000  1925.000000  1925.000000         1920.000000   
mean                 192.000000     0.467532     0.368831            0.108333   
std                  111.168431     0.499074     0.482613            0.310882   
min                    0.000000     0.000000     0.000000            0.000000   
25%                   96.000000     0.000000     0.000000            0.000000   
50%                  192.000000     0.000000     0.000000            0.000000   
75%                  288.000000     1.000000     1.000000            0.000000   
max                  384.000000     1.000000     1.000000            1.000000   

       DISEASE GROUPING 2  DISEASE GROUPING 3  DISEASE GROUPING 4  \
count         1920.000000         1920.000000         1920.000000   
mean             0.028125            0.097917            0.019792   
std              0.165373            0.297279            0.139320   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE GROUPING 5  DISEASE GROUPING 6          HTN  ...  \
count         1920.000000         1920.000000  1920.000000  ...   
mean             0.128125            0.046875     0.213021  ...   
std              0.334316            0.211426     0.409549  ...   
min              0.000000            0.000000     0.000000  ...   
25%              0.000000            0.000000     0.000000  ...   
50%              0.000000            0.000000     0.000000  ...   
75%              0.000000            0.000000     0.000000  ...   
max              1.000000            1.000000     1.000000  ...   

       RESPIRATORY_RATE_DIFF  TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
count            1925.000000       1925.000000             1925.000000   
mean               -0.758319         -0.811824               -0.910237   
std                 0.433037          0.296595                0.255396   
min                -1.000000         -1.000000               -1.000000   
25%                -1.000000         -1.000000               -1.000000   
50%                -1.000000         -1.000000               -1.000000   
75%                -0.705882         -0.690476               -0.919192   
max                 1.000000          1.000000                1.000000   

       BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
count                       1925.000000                      1925.000000   
mean                          -0.817464                        -0.758728   
std                            0.301522                         0.389168   
min                           -1.000000                        -1.000000   
25%                           -1.000000                        -1.000000   
50%                           -1.000000                        -1.000000   
75%                           -0.704088                        -0.613423   
max                            1.000000                         1.000000   

       HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
count          1925.000000                1925.000000           1925.000000   
mean             -0.848568                  -0.771306             -0.812740   
std               0.246814                   0.400915              0.295435   
min              -1.000000                  -1.000000             -1.000000   
25%              -1.000000                  -1.000000             -1.000000   
50%              -1.000000                  -1.000000             -1.000000   
75%              -0.764790                  -0.732494             -0.692177   
max               1.000000                   1.000000              

In [403]:
cleaned_df_patients = patients_filled_continuous_vars.dropna()

In [404]:
cleaned_df_patients

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                            0            1          60th       0   
1                            0            1          60th       0   
2                            0            1          60th       0   
3                            0            1          60th       0   
4                            0            1          60th       0   
...                        ...          ...           ...     ...   
1920                       384            0          50th       1   
1921                       384            0          50th       1   
1922                       384            0          50th       1   
1923                       384            0          50th       1   
1924                       384            0          50th       1   

      DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                    0.0                 0.0                 0.0   
1                    0.0                 0.0                 0.0   
2                    0.0                 0.0                 0.0   
3                    0.0                 0.0                 0.0   
4                    0.0                 0.0                 0.0   
...                  ...                 ...                 ...   
1920                 0.0                 0.0                 0.0   
1921                 0.0                 0.0                 0.0   
1922                 0.0                 0.0                 0.0   
1923                 0.0                 0.0                 0.0   
1924                 0.0                 0.0                 1.0   

      DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                    0.0                 1.0                 1.0  ...   
1                    0.0                 1.0                 1.0  ...   
2                    0.0                 1.0                 1.0  ...   
3                    0.0                 1.0                 1.0  ...   
4                    0.0                 1.0                 1.0  ...   
...                  ...                 ...                 ...  ...   
1920                 0.0                 0.0                 0.0  ...   
1921                 0.0                 0.0                 0.0  ...   
1922                 0.0                 0.0                 0.0  ...   
1923                 0.0                 0.0                 0.0  ...   
1924                 0.0                 0.0                 0.0  ...   

      TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
0            -1.000000               -1.000000   
1            -1.000000               -1.000000   
2            -1.000000               -1.000000   
3            -1.000000               -1.000000   
4            -0.238095               -0.818182   
...                ...                     ...   
1920         -1.000000               -1.000000   
1921         -1.000000               -1.000000   
1922         -1.000000               -1.000000   
1923         -1.000000               -1.000000   
1924         -0.547619               -0.838384   

      BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
0                            -1.000000                        -1.000000   
1                            -1.000000                        -1.000000   
2                            -0.389967                         0.407558   
3                            -0.389967                         0.407558   
4                            -0.389967                         0.407558   
...                                ...                              ...   
1920                         -1.000000                        -1.000000   
1921                         -1.000000                        -1.000000   
1922                         -1.000000                        -1.000000   
1923                         -1.000000                        -1.000000   
1924                         -0.701863                   

In [405]:
cleaned_df_patients.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65       GENDER  DISEASE GROUPING 1  \
count               1920.000000  1920.000000  1920.000000         1920.000000   
mean                 191.981771     0.468750     0.369792            0.108333   
std                  111.312587     0.499152     0.482874            0.310882   
min                    0.000000     0.000000     0.000000            0.000000   
25%                   95.750000     0.000000     0.000000            0.000000   
50%                  191.500000     0.000000     0.000000            0.000000   
75%                  288.250000     1.000000     1.000000            0.000000   
max                  384.000000     1.000000     1.000000            1.000000   

       DISEASE GROUPING 2  DISEASE GROUPING 3  DISEASE GROUPING 4  \
count         1920.000000         1920.000000         1920.000000   
mean             0.028125            0.097917            0.019792   
std              0.165373            0.297279            0.139320   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE GROUPING 5  DISEASE GROUPING 6          HTN  ...  \
count         1920.000000         1920.000000  1920.000000  ...   
mean             0.128125            0.046875     0.213021  ...   
std              0.334316            0.211426     0.409549  ...   
min              0.000000            0.000000     0.000000  ...   
25%              0.000000            0.000000     0.000000  ...   
50%              0.000000            0.000000     0.000000  ...   
75%              0.000000            0.000000     0.000000  ...   
max              1.000000            1.000000     1.000000  ...   

       RESPIRATORY_RATE_DIFF  TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
count            1920.000000       1920.000000             1920.000000   
mean               -0.758150         -0.812388               -0.910214   
std                 0.433588          0.296775                0.255728   
min                -1.000000         -1.000000               -1.000000   
25%                -1.000000         -1.000000               -1.000000   
50%                -1.000000         -1.000000               -1.000000   
75%                -0.705882         -0.690476               -0.919192   
max                 1.000000          1.000000                1.000000   

       BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
count                       1920.000000                      1920.000000   
mean                          -0.818008                        -0.759122   
std                            0.301726                         0.389598   
min                           -1.000000                        -1.000000   
25%                           -1.000000                        -1.000000   
50%                           -1.000000                        -1.000000   
75%                           -0.710805                        -0.622483   
max                            1.000000                         1.000000   

       HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
count          1920.000000                1920.000000           1920.000000   
mean             -0.848720                  -0.771187             -0.813295   
std               0.247117                   0.401430              0.295619   
min              -1.000000                  -1.000000             -1.000000   
25%              -1.000000                  -1.000000             -1.000000   
50%              -1.000000                  -1.000000             -1.000000   
75%              -0.764487                  -0.725806             -0.698390   
max               1.000000                   1.000000              

In [406]:
UCI_window_2_4_patients = (cleaned_df_patients
                               .query("WINDOW == '0-2' & ICU == 1")
                               ['PATIENT_VISIT_IDENTIFIER']
                               .values)

UCI_window_2_4_patients

array([  1,  17,  27,  41,  47,  48,  54,  62,  85,  90, 104, 106, 109,
       117, 144, 176, 202, 206, 226, 233, 239, 261, 270, 282, 286, 306,
       315, 329, 359, 362, 366, 379])

In [407]:
cleaned_df_patients = \
    cleaned_df_patients.query('PATIENT_VISIT_IDENTIFIER not in @UCI_window_2_4_patients')

cleaned_df_patients.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65       GENDER  DISEASE GROUPING 1  \
count               1760.000000  1760.000000  1760.000000          1760.00000   
mean                 192.818182     0.457386     0.380682             0.11250   
std                  110.637724     0.498322     0.485692             0.31607   
min                    0.000000     0.000000     0.000000             0.00000   
25%                   97.750000     0.000000     0.000000             0.00000   
50%                  191.500000     0.000000     0.000000             0.00000   
75%                  289.250000     1.000000     1.000000             0.00000   
max                  384.000000     1.000000     1.000000             1.00000   

       DISEASE GROUPING 2  DISEASE GROUPING 3  DISEASE GROUPING 4  \
count         1760.000000         1760.000000         1760.000000   
mean             0.026136            0.101705            0.021591   
std              0.159586            0.302345            0.145385   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE GROUPING 5  DISEASE GROUPING 6          HTN  ...  \
count         1760.000000         1760.000000  1760.000000  ...   
mean             0.132955            0.049432     0.212500  ...   
std              0.339622            0.216829     0.409193  ...   
min              0.000000            0.000000     0.000000  ...   
25%              0.000000            0.000000     0.000000  ...   
50%              0.000000            0.000000     0.000000  ...   
75%              0.000000            0.000000     0.000000  ...   
max              1.000000            1.000000     1.000000  ...   

       RESPIRATORY_RATE_DIFF  TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
count            1760.000000       1760.000000             1760.000000   
mean               -0.778209         -0.823458               -0.915593   
std                 0.411909          0.276279                0.247695   
min                -1.000000         -1.000000               -1.000000   
25%                -1.000000         -1.000000               -1.000000   
50%                -1.000000         -1.000000               -1.000000   
75%                -0.764706         -0.738095               -0.919192   
max                 1.000000          0.333333                1.000000   

       BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
count                       1760.000000                      1760.000000   
mean                          -0.828593                        -0.774857   
std                            0.287696                         0.368311   
min                           -1.000000                        -1.000000   
25%                           -1.000000                        -1.000000   
50%                           -1.000000                        -1.000000   
75%                           -0.726087                        -0.651942   
max                            1.000000                         1.000000   

       HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
count          1760.000000                1760.000000           1760.000000   
mean             -0.854690                  -0.788155             -0.824412   
std               0.240338                   0.384405              0.274953   
min              -1.000000                  -1.000000             -1.000000   
25%              -1.000000                  -1.000000             -1.000000   
50%              -1.000000                  -1.000000             -1.000000   
75%              -0.771625                  -0.756272             -0.736640   
max               1.000000                   1.000000              

we decided get the first line - window 0-2 - of each patient.
If that patient went to the UCI - future lines UCI == 1 - we will mark in the first line that the patient went to the UCI.


Pay attention note: we did bfill and ffill. Maybe we filled some information after the patient goes to the UCI. So we dropped when 0-2 window and UCI == 1, but maybe we filled the next lines of this patient with 0-2 window line information.
(I need reflect more on this...)

**Challenge 01: Do not use bfill in the window after the patient goes to the UCI.** 

Maybe use the function below first to do bfill and ffill is an alternative to solve the problem...

In [408]:
def window_prepare(rows):
    # check if that patient went to UCI, if yes
    if(np.any(rows['ICU'])):
        # attribute 1 for ICU in the row which the window is 0-2
        rows.loc[rows.WINDOW == '0-2', 'ICU'] = 1
        
    return rows.loc[rows.WINDOW == '0-2']



cleaned_df_patients = \
    cleaned_df_patients.groupby('PATIENT_VISIT_IDENTIFIER').apply(window_prepare)


cleaned_df_patients.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  \
PATIENT_VISIT_IDENTIFIER                                             
0                        0                          0            1   
2                        10                         2            0   
3                        15                         3            0   
4                        20                         4            0   
5                        25                         5            0   

                            AGE_PERCENTIL  GENDER  DISEASE GROUPING 1  \
PATIENT_VISIT_IDENTIFIER                                                
0                        0           60th       0                 0.0   
2                        10          10th       0                 0.0   
3                        15          40th       1                 0.0   
4                        20          10th       0                 0.0   
5                        25          10th       0                 0.0   

                             DISEASE GROUPING 2  DISEASE GROUPING 3  \
PATIENT_VISIT_IDENTIFIER                                              
0                        0                  0.0                 0.0   
2                        10                 0.0                 0.0   
3                        15                 0.0                 0.0   
4                        20                 0.0                 0.0   
5                        25                 0.0                 0.0   

                             DISEASE GROUPING 4  DISEASE GROUPING 5  \
PATIENT_VISIT_IDENTIFIER                                              
0                        0                  0.0                 1.0   
2                        10                 0.0                 0.0   
3                        15                 0.0                 0.0   
4                        20                 0.0                 0.0   
5                        25                 0.0                 0.0   

                             DISEASE GROUPING 6  ...  TEMPERATURE_DIFF  \
PATIENT_VISIT_IDENTIFIER                         ...                     
0                        0                  1.0  ...         -1.000000   
2                        10                 0.0  ...         -1.000000   
3                        15                 0.0  ...         -1.000000   
4                        20                 0.0  ...         -0.952381   
5                        25                 0.0  ...         -0.976190   

                             OXYGEN_SATURATION_DIFF  \
PATIENT_VISIT_IDENTIFIER                              
0                        0                -1.000000   
2                        10               -0.959596   
3                        15               -1.000000   
4                        20               -0.979798   
5                        25               -0.979798   

                             BLOODPRESSURE_DIASTOLIC_DIFF_REL  \
PATIENT_VISIT_IDENTIFIER                                        
0                        0                          -1.000000   
2                        10                         -0.515528   
3                        15                         -1.000000   
4                        20                         -1.000000   
5                        25                         -0.860870   

                             BLOODPRESSURE_SISTOLIC_DIFF_REL  \
PATIENT_VISIT_IDENTIFIER                                       
0                        0                         -1.000000   
2                        10                        -0.351328   
3                        15                        -1.000000   
4                        20                        -0.883669   
5                        25                        -0.714460   

                             HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  \
PATIENT_VISIT_IDENTIFIER                                                      
0                        0             -1.000000              

In [409]:
cleaned_df_patients[cleaned_df_patients.ICU == 1]

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  \
PATIENT_VISIT_IDENTIFIER                                               
0                        0                            0            1   
2                        10                           2            0   
11                       55                          11            1   
13                       65                          13            0   
14                       70                          14            1   
...                                                 ...          ...   
372                      1860                       372            1   
373                      1865                       373            1   
376                      1880                       376            1   
380                      1900                       380            0   
382                      1910                       382            0   

                              AGE_PERCENTIL  GENDER  DISEASE GROUPING 1  \
PATIENT_VISIT_IDENTIFIER                                                  
0                        0             60th       0                 0.0   
2                        10            10th       0                 0.0   
11                       55            80th       0                 0.0   
13                       65            50th       0                 0.0   
14                       70            60th       1                 0.0   
...                                     ...     ...                 ...   
372                      1860          80th       0                 0.0   
373                      1865          90th       1                 0.0   
376                      1880          90th       0                 1.0   
380                      1900          40th       1                 0.0   
382                      1910          50th       0                 0.0   

                               DISEASE GROUPING 2  DISEASE GROUPING 3  \
PATIENT_VISIT_IDENTIFIER                                                
0                        0                    0.0                 0.0   
2                        10                   0.0                 0.0   
11                       55                   0.0                 0.0   
13                       65                   0.0                 0.0   
14                       70                   0.0                 0.0   
...                                           ...                 ...   
372                      1860                 0.0                 0.0   
373                      1865                 0.0                 0.0   
376                      1880                 1.0                 1.0   
380                      1900                 0.0                 0.0   
382                      1910                 0.0                 0.0   

                               DISEASE GROUPING 4  DISEASE GROUPING 5  \
PATIENT_VISIT_IDENTIFIER                                                
0                        0                    0.0                 1.0   
2                        10                   0.0                 0.0   
11                       55                   0.0                 0.0   
13                       65                   0.0                 0.0   
14                       70                   0.0                 0.0   
...                                           ...                 ...   
372                      1860                 0.0                 0.0   
373                      1865                 0.0                 0.0   
376                      1880                 0.0                 1.0   
380                      1900                 0.0                 0.0   
382                      1910                 0.0                 0.0   

                               DISEASE GROUPING 6  ...  TEMPERATURE_DIFF  \
PATIENT_VISIT_IDENTIFIER                           ...                     
0                        0                    1.0  ...         -1.

In [410]:
len(cleaned_df_patients)

352

In [411]:
y = cleaned_df_patients['ICU']
x = cleaned_df_patients[x_columns].drop('ICU', axis=1)

In [412]:
np.random.seed(73246)

In [413]:
x_train, x_test, y_train, y_test = train_test_split(x, y, stratify=y)

In [414]:
dummy_model = DummyClassifier()
dummy_model.fit(x_train, y_train)

y_prediction_d = dummy_model.predict(x_test)
accuracy_score(y_test, y_prediction_d)

/home/gabriela/.local/lib/python3.6/site-packages/sklearn/dummy.py:132: FutureWarning: The default value of strategy will change from stratified to prior in 0.24.
  "stratified to prior in 0.24.", FutureWarning)


0.48863636363636365

In [415]:
LR_model = LogisticRegression()
LR_model.fit(x_train, y_train)

y_prediction = LR_model.predict(x_test)
accuracy_score(y_test, y_prediction)

/home/gabriela/.local/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.7045454545454546

**Challenge 02: Drop unnecessary columns.**

**Challenge 03: Review all the lessons.**

### Challenge 01: Challenge 01: Do not use bfill in the window after the patient goes to the UCI.

In [514]:
def ffill_df_continuous_vars(df: pd.DataFrame):
    
    continuous_features_cols = df.iloc[:, 13:-2].columns

    ffilled_continuous_features = (df.groupby('PATIENT_VISIT_IDENTIFIER')
                                   [continuous_features_cols]
                                   .fillna(method='ffill')
                                   .reset_index())
    
    categorical_features = df.iloc[:, :13]
    
    output = df.iloc[:, -2:]
    
    return pd.concat([categorical_features, ffilled_continuous_features, output], axis=1)

In [515]:
patients_ffilled_continuous_vars = ffill_df_continuous_vars(patients)

In [516]:
UCI_window_2_4_patients = (patients_ffilled_continuous_vars
                               .query("WINDOW == '0-2' & ICU == 1")
                               ['PATIENT_VISIT_IDENTIFIER']
                               .values)

cleaned_df_patients_2 = \
    patients_ffilled_continuous_vars \
        .query('PATIENT_VISIT_IDENTIFIER not in @UCI_window_2_4_patients') \
        .reset_index(drop=True)

cleaned_df_patients_2.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                         0            1          60th       0   
1                         0            1          60th       0   
2                         0            1          60th       0   
3                         0            1          60th       0   
4                         0            1          60th       0   

   DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                 0.0                 0.0                 0.0   
1                 0.0                 0.0                 0.0   
2                 0.0                 0.0                 0.0   
3                 0.0                 0.0                 0.0   
4                 0.0                 0.0                 0.0   

   DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                 0.0                 1.0                 1.0  ...   
1                 0.0                 1.0                 1.0  ...   
2                 0.0                 1.0                 1.0  ...   
3                 0.0                 1.0                 1.0  ...   
4                 0.0                 1.0                 1.0  ...   

   TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  BLOODPRESSURE_DIASTOLIC_DIFF_REL  \
0         -1.000000               -1.000000                         -1.000000   
1         -1.000000               -1.000000                         -1.000000   
2         -1.000000               -1.000000                         -1.000000   
3         -1.000000               -1.000000                         -1.000000   
4         -0.238095               -0.818182                         -0.389967   

   BLOODPRESSURE_SISTOLIC_DIFF_REL  HEART_RATE_DIFF_REL  \
0                        -1.000000            -1.000000   
1                        -1.000000            -1.000000   
2                        -1.000000            -1.000000   
3                        -1.000000            -1.000000   
4                         0.407558            -0.230462   

   RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
0                  -1.000000             -1.000000   
1                  -1.000000             -1.000000   
2                  -1.000000             -1.000000   
3                  -1.000000             -1.000000   
4                   0.096774             -0.242282   

   OXYGEN_SATURATION_DIFF_REL    WINDOW  ICU  
0                   -1.000000       0-2    0  
1                   -1.000000       2-4    0  
2                   -1.000000       4-6    0  
3                   -1.000000      6-12    0  
4                   -0.814433  ABOVE_12    1  

[5 rows x 232 columns]

In [517]:
def window_prepare_2(rows):
    # check if that patient went to UCI, if yes
    if(np.any(rows['ICU'])):
        # attribute 1 for ICU in the row which the window is 0-2
        rows.loc[rows.WINDOW == '0-2', 'ICU'] = 1
        
    return rows


patients_window_prepared_2 = cleaned_df_patients_2 \
                                 .groupby('PATIENT_VISIT_IDENTIFIER') \
                                 .apply(window_prepare_2) \
                                 .reset_index(drop=True)

In [518]:
patients_window_prepared_2.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                         0            1          60th       0   
1                         0            1          60th       0   
2                         0            1          60th       0   
3                         0            1          60th       0   
4                         0            1          60th       0   

   DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                 0.0                 0.0                 0.0   
1                 0.0                 0.0                 0.0   
2                 0.0                 0.0                 0.0   
3                 0.0                 0.0                 0.0   
4                 0.0                 0.0                 0.0   

   DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                 0.0                 1.0                 1.0  ...   
1                 0.0                 1.0                 1.0  ...   
2                 0.0                 1.0                 1.0  ...   
3                 0.0                 1.0                 1.0  ...   
4                 0.0                 1.0                 1.0  ...   

   TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  BLOODPRESSURE_DIASTOLIC_DIFF_REL  \
0         -1.000000               -1.000000                         -1.000000   
1         -1.000000               -1.000000                         -1.000000   
2         -1.000000               -1.000000                         -1.000000   
3         -1.000000               -1.000000                         -1.000000   
4         -0.238095               -0.818182                         -0.389967   

   BLOODPRESSURE_SISTOLIC_DIFF_REL  HEART_RATE_DIFF_REL  \
0                        -1.000000            -1.000000   
1                        -1.000000            -1.000000   
2                        -1.000000            -1.000000   
3                        -1.000000            -1.000000   
4                         0.407558            -0.230462   

   RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
0                  -1.000000             -1.000000   
1                  -1.000000             -1.000000   
2                  -1.000000             -1.000000   
3                  -1.000000             -1.000000   
4                   0.096774             -0.242282   

   OXYGEN_SATURATION_DIFF_REL    WINDOW  ICU  
0                   -1.000000       0-2    1  
1                   -1.000000       2-4    0  
2                   -1.000000       4-6    0  
3                   -1.000000      6-12    0  
4                   -0.814433  ABOVE_12    1  

[5 rows x 232 columns]

In [519]:
def bfill_df_continuous_vars(df: pd.DataFrame):
    
    continuous_features_cols = df.iloc[:, 13:-2].columns

    bfilled_continuous_features = (df.groupby('PATIENT_VISIT_IDENTIFIER')
                                   [continuous_features_cols]
                                   .fillna(method='bfill')
                                   .reset_index())
    
    categorical_features = df.iloc[:, :13]
    
    output = df.iloc[:, -2:]
    
    return pd.concat([categorical_features, bfilled_continuous_features, output], axis=1)

In [520]:
patients_bffilled_continuous_vars = bfill_df_continuous_vars(patients_window_prepared_2)

patients_bffilled_continuous_vars

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                            0            1          60th       0   
1                            0            1          60th       0   
2                            0            1          60th       0   
3                            0            1          60th       0   
4                            0            1          60th       0   
...                        ...          ...           ...     ...   
1760                       384            0          50th       1   
1761                       384            0          50th       1   
1762                       384            0          50th       1   
1763                       384            0          50th       1   
1764                       384            0          50th       1   

      DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                    0.0                 0.0                 0.0   
1                    0.0                 0.0                 0.0   
2                    0.0                 0.0                 0.0   
3                    0.0                 0.0                 0.0   
4                    0.0                 0.0                 0.0   
...                  ...                 ...                 ...   
1760                 0.0                 0.0                 0.0   
1761                 0.0                 0.0                 0.0   
1762                 0.0                 0.0                 0.0   
1763                 0.0                 0.0                 0.0   
1764                 0.0                 0.0                 1.0   

      DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                    0.0                 1.0                 1.0  ...   
1                    0.0                 1.0                 1.0  ...   
2                    0.0                 1.0                 1.0  ...   
3                    0.0                 1.0                 1.0  ...   
4                    0.0                 1.0                 1.0  ...   
...                  ...                 ...                 ...  ...   
1760                 0.0                 0.0                 0.0  ...   
1761                 0.0                 0.0                 0.0  ...   
1762                 0.0                 0.0                 0.0  ...   
1763                 0.0                 0.0                 0.0  ...   
1764                 0.0                 0.0                 0.0  ...   

      TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
0            -1.000000               -1.000000   
1            -1.000000               -1.000000   
2            -1.000000               -1.000000   
3            -1.000000               -1.000000   
4            -0.238095               -0.818182   
...                ...                     ...   
1760         -1.000000               -1.000000   
1761         -1.000000               -1.000000   
1762         -1.000000               -1.000000   
1763         -1.000000               -1.000000   
1764         -0.547619               -0.838384   

      BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
0                            -1.000000                        -1.000000   
1                            -1.000000                        -1.000000   
2                            -1.000000                        -1.000000   
3                            -1.000000                        -1.000000   
4                            -0.389967                         0.407558   
...                                ...                              ...   
1760                         -1.000000                        -1.000000   
1761                         -1.000000                        -1.000000   
1762                         -1.000000                        -1.000000   
1763                         -1.000000                        -1.000000   
1764                         -0.701863                   

In [521]:
cleaned_df_patients_2 = patients_bffilled_continuous_vars.dropna()

len(cleaned_df_patients_2)

1755

In [522]:
cleaned_df_patients_2 = cleaned_df_patients_2[cleaned_df_patients_2.WINDOW == '0-2']

In [523]:
cleaned_df_patients_2.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65      GENDER  DISEASE GROUPING 1  \
count                351.000000   351.000000  351.000000          351.000000   
mean                 192.549858     0.458689    0.381766            0.105413   
std                  110.807231     0.499002    0.486513            0.307523   
min                    0.000000     0.000000    0.000000            0.000000   
25%                   97.500000     0.000000    0.000000            0.000000   
50%                  191.000000     0.000000    0.000000            0.000000   
75%                  289.500000     1.000000    1.000000            0.000000   
max                  384.000000     1.000000    1.000000            1.000000   

       DISEASE GROUPING 2  DISEASE GROUPING 3  DISEASE GROUPING 4  \
count          351.000000          351.000000          351.000000   
mean             0.022792            0.091168            0.019943   
std              0.149453            0.288259            0.140004   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE GROUPING 5  DISEASE GROUPING 6         HTN  ...  \
count          351.000000          351.000000  351.000000  ...   
mean             0.128205            0.042735    0.193732  ...   
std              0.334795            0.202548    0.395786  ...   
min              0.000000            0.000000    0.000000  ...   
25%              0.000000            0.000000    0.000000  ...   
50%              0.000000            0.000000    0.000000  ...   
75%              0.000000            0.000000    0.000000  ...   
max              1.000000            1.000000    1.000000  ...   

       RESPIRATORY_RATE_DIFF  TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
count             351.000000        351.000000              351.000000   
mean               -0.961957         -0.965134               -0.985323   
std                 0.110550          0.086235                0.032677   
min                -1.000000         -1.000000               -1.000000   
25%                -1.000000         -1.000000               -1.000000   
50%                -1.000000         -1.000000               -1.000000   
75%                -1.000000         -1.000000               -0.979798   
max                -0.058824         -0.238095               -0.757576   

       BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
count                        351.000000                       351.000000   
mean                          -0.962986                        -0.958804   
std                            0.106265                         0.117708   
min                           -1.000000                        -1.000000   
25%                           -1.000000                        -1.000000   
50%                           -1.000000                        -1.000000   
75%                           -1.000000                        -1.000000   
max                            0.130435                         0.378657   

       HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
count           351.000000                 351.000000            351.000000   
mean             -0.976353                  -0.964289             -0.965628   
std               0.059678                   0.095232              0.084406   
min              -1.000000                  -1.000000             -1.000000   
25%              -1.000000                  -1.000000             -1.000000   
50%              -1.000000                  -1.000000             -1.000000   
75%              -1.000000                  -1.000000             -1.000000   
max              -0.535878                  -0.268817             -0.285468   

     

In [524]:
y = cleaned_df_patients_2['ICU']
x = cleaned_df_patients_2[x_columns].drop('ICU', axis=1)

In [525]:
np.random.seed(73246)

In [526]:
x_train, x_test, y_train, y_test = train_test_split(x, y, stratify=y)

len(x_train)

263

In [527]:
dummy_model = DummyClassifier()
dummy_model.fit(x_train, y_train)

y_prediction_d = dummy_model.predict(x_test)
accuracy_score(y_test, y_prediction_d)

/home/gabriela/.local/lib/python3.6/site-packages/sklearn/dummy.py:132: FutureWarning: The default value of strategy will change from stratified to prior in 0.24.
  "stratified to prior in 0.24.", FutureWarning)


0.48863636363636365

In [528]:
LR_model = LogisticRegression()
LR_model.fit(x_train, y_train)

y_prediction = LR_model.predict(x_test)
accuracy_score(y_test, y_prediction)

/home/gabriela/.local/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.7045454545454546

It doesnt change the result.

### Challenge 02: Drop unnecessary columns.

In [555]:
list(cleaned_df_patients_2.columns.values)
x_columns.values

array(['PATIENT_VISIT_IDENTIFIER', 'AGE_ABOVE65', 'GENDER', 'DISEASE GROUPING 1', 'DISEASE GROUPING 2', 'DISEASE GROUPING 3', 'DISEASE GROUPING 4', 'DISEASE GROUPING 5', 'DISEASE GROUPING 6', 'HTN',
       'IMMUNOCOMPROMISED', 'OTHER', 'ALBUMIN_MEDIAN', 'ALBUMIN_MEAN', 'ALBUMIN_MIN', 'ALBUMIN_MAX', 'ALBUMIN_DIFF', 'BE_ARTERIAL_MEDIAN', 'BE_ARTERIAL_MEAN', 'BE_ARTERIAL_MIN', 'BE_ARTERIAL_MAX',
       'BE_ARTERIAL_DIFF', 'BE_VENOUS_MEDIAN', 'BE_VENOUS_MEAN', 'BE_VENOUS_MIN', 'BE_VENOUS_MAX', 'BE_VENOUS_DIFF', 'BIC_ARTERIAL_MEDIAN', 'BIC_ARTERIAL_MEAN', 'BIC_ARTERIAL_MIN',
       'BIC_ARTERIAL_MAX', 'BIC_ARTERIAL_DIFF', 'BIC_VENOUS_MEDIAN', 'BIC_VENOUS_MEAN', 'BIC_VENOUS_MIN', 'BIC_VENOUS_MAX', 'BIC_VENOUS_DIFF', 'BILLIRUBIN_MEDIAN', 'BILLIRUBIN_MEAN',
       'BILLIRUBIN_MIN', 'BILLIRUBIN_MAX', 'BILLIRUBIN_DIFF', 'BLAST_MEDIAN', 'BLAST_MEAN', 'BLAST_MIN', 'BLAST_MAX', 'BLAST_DIFF', 'CALCIUM_MEDIAN', 'CALCIUM_MEAN', 'CALCIUM_MIN', 'CALCIUM_MAX',
       'CALCIUM_DIFF', 'CREATININ_MEDIAN', '

- Let's try use just the disease columns

In [556]:
x = cleaned_df_patients_2[x_columns].iloc[:, :12]
y = cleaned_df_patients_2['ICU']

In [557]:
np.random.seed(73246)

In [558]:
x_train, x_test, y_train, y_test = train_test_split(x, y, stratify=y)

len(x_train)

263

In [559]:
dummy_model = DummyClassifier()
dummy_model.fit(x_train, y_train)

y_prediction_d = dummy_model.predict(x_test)
accuracy_score(y_test, y_prediction_d)

/home/gabriela/.local/lib/python3.6/site-packages/sklearn/dummy.py:132: FutureWarning: The default value of strategy will change from stratified to prior in 0.24.
  "stratified to prior in 0.24.", FutureWarning)


0.48863636363636365

In [560]:
LR_model = LogisticRegression()
LR_model.fit(x_train, y_train)

y_prediction = LR_model.predict(x_test)
accuracy_score(y_test, y_prediction)

/home/gabriela/.local/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.5568181818181818

- It got worse.

### Challenge 03: Review all the lessons.

checked.